src/common/utils/registry/auth/tokenauthorizer.go

```diff
@@ -40,209 +44,308 @@ func (s *Scope) string() string {
+		}
+		token = t.getCachedToken(key)
+	}
+
+	// request a new token if the token is null
+	if token == nil {
+		token, err = t.generator.generate(scopes, t.registryURL.String())
+		if err != nil {
+			return err
+		}
+		if token == nil {
+			return nil
+		}
+		// only cache the token for empty scope(login) or single scope request
+		if len(scopes) <= 1 {
+			key := ""
+			if len(scopes) == 1 {
+				key = scopes[0].string()
+			}
+			t.updateCachedToken(key, token)
+		}
+	}
+
+	req.Header.Add(http.CanonicalHeaderKey("Authorization"), fmt.Sprintf("Bearer %s", token.Token))
+
+	return nil
 }
 
-// AuthorizeRequest will add authorization header which contains a token before the request is sent
-func (t *tokenAuthorizer) Authorize(req *http.Request, params map[string]string) error {
-	var scopes []*Scope
-	var token string
+// some requests are sent to backend storage, such as s3, this method filters
+// the requests only sent to registry
+func (t *tokenAuthorizer) filterReq(req *http.Request) (bool, error) {
+	// the registryURL is nil when the first request comes, init it with
+	// the scheme and host of the request which must be sent to the registry
+	if t.registryURL == nil {
+		u, err := url.Parse(buildPingURL(req.URL.Scheme + "://" + req.URL.Host))
+		if err != nil {
+			return false, err
+		}
+		t.registryURL = u
+	}
+
+	v2Index := strings.Index(req.URL.Path, "/v2/")
+	if v2Index == -1 {
+		return false, nil
+	}
+
+	if req.URL.Host != t.registryURL.Host || req.URL.Scheme != t.registryURL.Scheme ||
+		req.URL.Path[:v2Index+4] != t.registryURL.Path {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// parse scopes from the request according to its method, path and query string
+func parseScopes(req *http.Request) []*Scope {
+	scopes := []*Scope{}
 
-	hasFrom := false
 	from := req.URL.Query().Get("from")
 	if len(from) != 0 {
-		s := &Scope{
+		scopes = append(scopes, &Scope{
 			Type:    "repository",
 			Name:    from,
 			Actions: []string{"pull"},
+		})
+	}
+
+	var scope *Scope
+	path := strings.TrimRight(req.URL.Path, "/")
+	repository := parseRepository(path)
+	if len(repository) > 0 {
+		// pull, push, delete blob/manifest
+		scope = &Scope{
+			Type: "repository",
+			Name: repository,
+		}
+		switch req.Method {
+		case http.MethodGet:
+			scope.Actions = []string{"pull"}
+		case http.MethodPost, http.MethodPut, http.MethodPatch:
+			scope.Actions = []string{"push"}
+		case http.MethodDelete:
+			scope.Actions = []string{"*"}
+		default:
+			scope = nil
+			log.Warningf("unsupported method: %s", req.Method)
+		}
+	} else if catalog.MatchString(path) {
+		// catalog
+		scope = &Scope{
+			Type:    "registry",
+			Name:    "catalog",
+			Actions: []string{"*"},
 		}
-		scopes = append(scopes, s)
-		// do not cache the token if "from" appears
-		hasFrom = true
+	} else if base.MatchString(path) {
+		// base
+		scope = nil
+	} else {
+		// unknow
+		log.Warningf("can not parse scope from the request: %s %s", req.Method, req.URL.Path)
```

Comment:
must handle not match condition, otherwise all the non-v2 request will get the login token